### PR TITLE
RUST-1069 Defer implicit session check out until after connection check out

### PR DESCRIPTION
**Pull request by abr-egn**
RUST-1069

Previously, retries were handled by a distinct `execute_retry` method; this was an issue for this change because there are two retry cases (connection check out failure and operation failure), and a retry should create an implicit session in the first case but not the second.  This PR changes it to have a loop inside `execute_operation_with_retry`; this makes the handling there a bit more complicated, but removes the `execute_retry` method entirely, and also collapses `ExecutionOutput` into `ExecutionDetails`, so I think it's a net wash on complexity.  Having the loop will also make executing multiple retries easier if/when we get to CSOT.

### CODE REVIEW

File: src/client/executor.rs
Comment (isabelatkinson) [inline src/client/executor.rs:357]: Is there any chance this could fail during a retry? If so, I think we should return the original error rather than the one that occurs in this call.

File: src/client/executor.rs
Comment (isabelatkinson) [inline src/client/executor.rs:298]: Can this be moved outside of the loop? I don't think these conditions would change on a retry: the session and operation should each stay the same.

File: src/client/executor.rs
Comment (isabelatkinson) [inline src/client/executor.rs:354]: IIUC here an implicit session will be started only on the first attempt and reused in subsequent retries. Is this the right behavior? I'm not sure if we should be dropping/re-creating the implicit session after checking out a connection each time around.

File: src/client/executor.rs
Comment (isabelatkinson) [inline src/client/executor.rs:313]: ditto above, does this need to be done in the loop?

**Review by isabelatkinson — COMMENTED** [review]
Just a few questions. As I reviewed this I noticed some areas in this file that could use refactoring, but I don't think they're relevant to this PR so I filed RUST-1575 to address them in the future.

File: src/client/executor.rs
Comment (abr-egn) [reply src/client/executor.rs]: Yup, done.

File: src/client/executor.rs
Comment (abr-egn) [reply src/client/executor.rs:313]: Unfortunately not - this (potentially) borrows from `op`, so moving it outside the loop would block the call to `op.update_for_retry()` (which requires a &mut).

File: src/client/executor.rs
Comment (abr-egn) [reply src/client/executor.rs:354]: I believe this is correct - the goal of the upstream change was to minimize the number of sessions created serverside.

File: src/client/executor.rs
Comment (abr-egn) [reply src/client/executor.rs:357]: The only error this can return is an internal one indicating an impossible/invalid connection state, which I think would be valuable to expose if it does happen :)

**Review by isabelatkinson — APPROVED** [review]

**Comment by isabelatkinson** [comment]
not sure if we want another review on this -- I'll leave it up to you, Abraham, if you want to wait until Kaitlin is back

**Comment by abr-egn** [comment]
> not sure if we want another review on this -- I'll leave it up to you, Abraham, if you want to wait until Kaitlin is back

No rush. and this is core enough that more eyes will be good :)

File: src/client/executor.rs
Comment (kmahar) [reply src/client/executor.rs:354]: the retryable writes spec actually _requires_ that the retry attempt use the same session as the initial operation, so this is definitely right, and I assume there are some spec tests that would start failing if we didn't preserve this behavior.

per [spec](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#executing-retryable-write-commands):

>  When retrying a write command, drivers MUST resend the command with the same transaction ID.

where "transaction ID" is [defined](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#terms) as:

> the top-level lsid and txnNumber fields.

and per the sessions [spec](https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#why-do-automatic-retry-attempts-re-use-a-dirty-implicit-session):

> The retryable writes spec requires that both the original and retry attempt use the same server session. The server will block the retry attempt until the initial attempt completes at which point the retry attempt will continue executing.
For retryable reads that use an implicit session, drivers could choose to use a new server session for the retry attempt however this would lose the information that these two reads are related.

**Review by kmahar — APPROVED** [review]
one comment that I don't need to re-review for, LGTM otherwise.

File: src/test/spec/sessions.rs
Comment (abr-egn) [reply src/test/spec/sessions.rs:166]: Good call, done.

**Pull request merged**
